### DISCUSSION TITLE
Optimize "to_world" in envmap and directional emitters for automatic HDR and sunlight placement in a VFX production

### DISCUSSION THREAD

**Discussion by ultra-sonic** [mitsuba-renderer/mitsuba3#1396] [Feature requests]
Hi there,

First off, I just want to say I’m a huge fan of your work on Mitsuba3—it’s absolutely phenomenal! My name is Oli, and I’m the Head of 3D at RISE. When I came across Mitsuba3, my immediate thought was, "Wow, let's optimize the 'to_world' parameter for envmaps and directional lights!"

Here at RISE, we often receive HDRIs from set, accompanied by shot-specific image sequences with gray and chrome balls. Automating the "to_world" parameter would be a game-changer for processing our refBall+HDRI packages efficiently.

After experimenting with Mitsuba3, I noticed that the "to_world" parameter isn’t currently differentiable. I was curious—are there plans to make this differentiable in the future? Or is there a specific reason why this hasn’t been implemented yet?

Thank you so much for all your amazing work on Mitsuba3. I’d love to hear your thoughts on this, and I’m happy to provide any additional context if needed!

Best regards,
Oli

**Comment by ultra-sonic** [comment]
i forgot: we have the correct camera position and we (roughly) know the position and size of the greyball.

**Reply by njroussel** [reply]
Hi @ultra-sonic 

This is an interesting application and I think that it is theoretically feasible with a few changes in the source code. However, I'm not sure practical it would be  - I imagine that a fairly good initialization would be required or else the optimization might find itself stuck in a local minimum that is quite far from the target. 

We're fairly busy with some deadlines at the moment, I'll try this myself in a couple of weeks :+1:

**Reply by njroussel** [reply]
Actually, I take that partially back. You should be able to do this right now. 

Rather than optimizing the `to_world` parameter, you can optimize the `data` parameter. Of course, you don't want it to reconstruct the pixels indivdually. So instead, you make use of a latent variable that controls the `data` parameter. Effectively this means that you're rotating the envmap in image space rather than in world space. You can have a look at how this is done in the [object pose estimation tutorial](https://mitsuba.readthedocs.io/en/stable/src/inverse_rendering/object_pose_estimation.html) where we optimize a transform matrix as a latent variable that controls the individual vertex positions of a triangle mesh.

**Reply by ultra-sonic** [reply]
Hi Nicolas,

thank you so much for gettiing back to me! Much appreciated!

I already thought about optimizing "data" as well. So i will give this a go for the HDR envmap.

Instead of using a directional emitter for the sun I will just extract it from the HDR into a sperate envmap and use the same rotation that i used for the HDR and optimize the intensity of the sun additionally, because HDRs usually do not store the real sun intensity (people on a film-set just do 3-7 exposure brackets that never capture the full intensity)

I will let you know how it went!
Thanks again
Oli

**Reply by AnsonSavage** [reply]
This is such a neat application! I'm curious, has this latent variable method worked out for you?

**Reply by danielkorth** [reply]
Would also be curious to hear. I'm trying to optimize both for the rotation and for the HDR map itself, and cannot quite get it to work
